### ISSUE TITLE
Failures in parallel tasks do not stop execution

### ISSUE THREAD

**Issue by sdcooke** [fabric/fabric#457]
The following test case runs `test.sh` which returns an error exit code. Without the `@parallel` decorator fabric aborts and exits as expected after the first failure on the first host (there are two hosts in the 'web_workers' role). With the `@parallel` decorator the `test_thing` task exits before running `date` but `test_thing2` is then run.

``` python
def test():
    execute(test_thing)
    execute(test_thing2)

@roles('web_workers')
@parallel
def test_thing():
    run('./test.sh')
    run('date')

@roles('web_workers')
@parallel
def test_thing2():
    run('date')
```

I'm assuming this isn't the intended workflow, if it is or I'm making a mistake I'd be interested to know what it is.

**Comment by paulRbr** [comment]
Hi guys,

I would like to talk again about this issue (Yeah I'm ⛏).

Here is my use case of the parallel paradigm of Fabric:
```python
@runs_once
def deploy
  execute(upload)
  execute(update)

@parallel
def upload
  # upload app artifact

@parallel(pool_size=amount_servers/4)
def update
  # stop and start app
```
called like the following:
```bash
> fab -R servers deploy
```

I would expect Fabric to stop the parallel execution of the `update` task to stop whenever a task exists with a non zero code. At least stop after the end of the current pool. But it does not and Fabric executes the `update` task on every host whatever the exit code is.

That's quite problematic if the app could not start in the `update` task as we end up stopping all instances of the app.

Do you have an idea on how to solve this?

**Comment by paulRbr** [comment]
ping @bitprophet @goosemo should I open a new issue with by comment above?

**Comment by jsalva** [comment]
Having a similar issue. +1 on a solution

**Comment by goosemo** [comment]
@paulRbr just now looking at this, sorry for the delay. But have you seen this PR? https://github.com/fabric/fabric/pull/1471

It's potentially solving your issue if I am understanding correctly.

**Comment by jsalva** [comment]
it turns out this doesn't fix my problem. My issue is that the instances are attached to a load balancer on AWS, and can be terminated at random. It seems as though the terminated instance does not cause failure unless the task finishes with an actual exit code. Fabric just hangs into perpetuity, waiting for the now non-existent host to finish.

**Comment by goosemo** [comment]
@jsalva interesting, are they being killed in the middle of a workload?

**Comment by paulRbr** [comment]
Hi @goosemo, oh that's a nice PR. It does exactly what we need @captaintrain! Thanks!

Here's the output without the PR:
```bash
> fab -H local1,local2 deploy
[local1] Executing task 'deploy'
[local1] Executing task 'upload'
[local2] Executing task 'upload'
[localhost] local: # hello upload
[localhost] local: # hello upload
[local1] Executing task 'update'
[local2] Executing task 'update'
[localhost] local: # hello update
[localhost] local: # hello update    < I DON'T want this second one to run as the first update fails

Fatal error: One or more hosts failed while executing task 'update'

Aborting.
```

And now the output with the PR and the `--parallel-exit-on-errors` flag:
```bash
> fab --parallel-exit-on-errors -H local1,local2 deploy
[local1] Executing task 'deploy'
[local1] Executing task 'upload'
[local2] Executing task 'upload'
[localhost] local: # hello upload
[localhost] local: # hello upload
[local1] Executing task 'update'
[local2] Executing task 'update'
[localhost] local: # hello update     < \o/ It ran only once!

Fatal error: One or more hosts failed while executing task 'update'

Aborting.
```

**Comment by jsalva** [comment]
@goosemo I'm intentionally terminating the hosts mid run()/sudo(), so mid-workload rather than stale hosts

**Comment by jsalva** [comment]
it looks like we could in principle detect the state where the transport.send_ignore() raises a ProxyCommandFailure  and use that to signal that the connection is closed, terminating the queued process with a failed status code of some sort.
